### PR TITLE
new module: riboWaltz

### LINKED ISSUE #5489: new module: riboWaltz
### Is there an existing module for this?

- [X] I have searched for the existing module

### Is there an open PR for this?

- [X] I have searched for existing PRs

### Is there an open issue for this?

- [X] I have searched for existing issues

### Are you going to work on this?

- [X] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### PR DESCRIPTION
**Pull request by iraiosub**
Add new module for the riboWaltz tool for identifying P-sites and performing diagnostics on ribo-seq data.
## PR checklist

Closes #5489  

- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - For modules:
    - [x] `nf-core modules test <MODULE> --profile docker`
    - [ ] `nf-core modules test <MODULE> --profile singularity`
    - [ ] `nf-core modules test <MODULE> --profile conda`

### CODE REVIEW

File: modules/nf-core/ribowaltz/templates/ribowaltz.r [hunk 1]
```diff
@@ -0,0 +1,479 @@
...
+
+# Calculation of optimal P-site offsets, diagnostic analysis and visual inspection of ribosome profiling data
+# Author: Ira Iosub
+# Example usage: Rscript ribowaltz.r --bam SRX11780887.Aligned.toTranscriptome.out.bam --gtf Homo_sapiens.GRCh38.111_chr20.gtf --fasta Homo_sapiens.GRCh38.dna.chromosome.20.fa.gz --length_range "26:34" --periodicity_threshold 50 --exclude_start 42 --exclude_stop 27
+
+suppressPackageStartupMessages(library(riboWaltz))
+suppressPackageStartupMessages(library(dplyr))
+
+# Allow data.table to use all available threads
+data.table::setDTthreads(0)
```
Comment (SPPearce) [inline modules/nf-core/ribowaltz/templates/ribowaltz.r:11]: all available threads? Shouldn't this be using `task.cpus`?

**Review by SPPearce — COMMENTED** [review]
I haven't gone through the whole R code, just the Nextflow parts. Is there someone who knows about riboseq who can review the R?

File: modules/nf-core/ribowaltz/meta.yml [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+output:
+  #Only when we have meta
+  - meta:
+      type: map
+      description: |
+        Groovy Map containing sample information
+        e.g. `[ id:'sample1', single_end:false ]`
+
+  - versions:
```
Comment (pinin4fjords) [inline modules/nf-core/ribowaltz/meta.yml:57]: Can you order these the same as in the main.nf for ease of comparison?

File: modules/nf-core/ribowaltz/templates/ribowaltz.r [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+    # Ensure the option vectors are length 2 (key/ value) to catch empty ones
+    args_vals <- lapply(args_vals, function(z){ length(z) <- 2; z})
+
+    parsed_args <- structure(lapply(args_vals, function(x) x[2]), names = lapply(args_vals, function(x) x[1]))
+    parsed_args[! is.na(parsed_args)]
+}
+
+
+# Function to print help menu
+print_help_menu <- function() {
```
Comment (pinin4fjords) [inline modules/nf-core/ribowaltz/templates/ribowaltz.r:53]: This is never going to be run interactively, I don't think the inline help makes sense to include.

File: modules/nf-core/ribowaltz/templates/ribowaltz.r [hunk 4]
```diff
@@ -0,0 +1,479 @@
+#!/usr/bin/env Rscript
+
+# Calculation of optimal P-site offsets, diagnostic analysis and visual inspection of ribosome profiling data
+# Author: Ira Iosub
+# Example usage: Rscript ribowaltz.r --bam SRX11780887.Aligned.toTranscriptome.out.bam --gtf Homo_sapiens.GRCh38.111_chr20.gtf --fasta Homo_sapiens.GRCh38.dna.chromosome.20.fa.gz --length_range "26:34" --periodicity_threshold 50 --exclude_start 42 --exclude_stop 27
```
Comment (pinin4fjords) [inline modules/nf-core/ribowaltz/templates/ribowaltz.r:5]: This is never going to be called this way as an nf-core template, so this should be removed.

File: modules/nf-core/ribowaltz/templates/ribowaltz.r [hunk 5]
```diff
@@ -0,0 +1,479 @@
...
+
+# Create folder for plots
+dir.create("ribowaltz_qc")
+
+
+# =========
+# Functions for riboWaltz analysis
+# =========
+
+export_offsets <- function(name, df) {
```
Comment (pinin4fjords) [inline modules/nf-core/ribowaltz/templates/ribowaltz.r:139]: Keep the functions in the same place. So move these up and apply function documentation for consistency.

**Review by pinin4fjords — CHANGES_REQUESTED** [review]
My first pass, in addition to Simon's points (he's right as usual).

This is not going to be called interactively, so you need to remove the vestiges of that in the template R file.

Keep all the functions together and properly document them all. 

Also try to be sure you're not layering too much of your own 'code creativity' on top of ribowaltz. If we're calling this the 'ribowaltz' module then users will be expecting this to do what Ribowaltz functions do, and not too much else.

File: modules/nf-core/ribowaltz/main.nf [hunk 6]
```diff
@@ -0,0 +1,48 @@
+process RIBOWALTZ {
+    tag "$meta.id"
+    label 'process_high'
```
[the remark this answers is not in the document]
Comment (iraiosub) [reply modules/nf-core/ribowaltz/main.nf]: process_medium should be ok for most users. see commit [2c16f8e](2c16f8e)

File: modules/nf-core/ribowaltz/main.nf [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    output:
+    tuple val(meta), path("*.psite_offset.tsv{,.gz}"), emit: offset, optional: true
+    tuple val(meta), path("offset_plot/*"), emit: offset_plot, optional: true
+    tuple val(meta), path("*.psite.tsv{,.gz}"), emit: psites, optional: true
+    tuple val(meta), path("*.codon_coverage_rpf.tsv{,.gz}"), emit: codon_coverage_rpf, optional: true
+    tuple val(meta), path("*.codon_coverage_psite.tsv{,.gz}"), emit: codon_coverage_psite, optional: true
+    tuple val(meta), path("*.cds_coverage_psite.tsv{,.gz}"), emit: cds_coverage, optional: true
+    tuple val(meta), path("*nt_coverage_psite.tsv{,.gz}"), emit: cds_window_coverage, optional: true
+    tuple val(meta), path("ribowaltz_qc/*.pdf"), emit: ribowaltz_qc, optional: true
+    path "versions.yml"           , emit: versions
```
[the remark this answers is not in the document]
Comment (iraiosub) [reply modules/nf-core/ribowaltz/main.nf]: done, see [2c16f8e](2c16f8e)

File: modules/nf-core/ribowaltz/meta.yml [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+  - "ribowaltz":
+      description: "Calculation of optimal P-site offsets, diagnostic analysis and visual inspection of ribosome profiling data."
+      homepage: "https://github.com/LabTranslationalArchitectomics/riboWaltz"
+      documentation: "https://github.com/LabTranslationalArchitectomics/riboWaltz"
+      tool_dev_url: "https://github.com/LabTranslationalArchitectomics/riboWaltz"
+      doi: "10.1371/journal.pcbi.1006169"
+      licence: ["MIT"]
+
+input:
+  # Only when we have meta
```
[the remark this answers is not in the document]
Comment (iraiosub) [reply modules/nf-core/ribowaltz/meta.yml]: thanks! done, see commit [2c16f8e](2c16f8e)

File: modules/nf-core/ribowaltz/meta.yml [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+      type: file
+      description: GTF file of reference genome
+      pattern: "*.{gtf.gz,gtf}"
+  - fasta:
+      type: file
+      description: FASTA file of reference genome
+      pattern: "*.{fasta,fa,fasta.gz,fa.gz}"
+
+output:
+  #Only when we have meta
```
[the remark this answers is not in the document]
Comment (iraiosub) [reply modules/nf-core/ribowaltz/meta.yml]: resolved in commit [2c16f8e](2c16f8e)

File: modules/nf-core/ribowaltz/meta.yml [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+      type: map
+      description: |
+        Map containing reference information for the reference genome GTF file
+        e.g. `[ id:'Ensembl human v.111' ]`
+  - meta3:
+      type: map
+      description: |
+        Map containing reference information for the reference genome FASTA file
+        e.g. `[ id:'Ensembl human v.111' ]`
+
```
[the remark this answers is not in the document]
Comment (iraiosub) [reply modules/nf-core/ribowaltz/meta.yml]: done. Referencing commit 2c16f8e.

File: modules/nf-core/ribowaltz/meta.yml [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+      pattern: "*.{fasta,fa,fasta.gz,fa.gz}"
+
+output:
+  #Only when we have meta
+  - meta:
+      type: map
+      description: |
+        Groovy Map containing sample information
+        e.g. `[ id:'sample1', single_end:false ]`
+
```
[the remark this answers is not in the document]
Comment (iraiosub) [reply modules/nf-core/ribowaltz/meta.yml]: done. Referencing commit 2c16f8e.

File: modules/nf-core/ribowaltz/templates/ribowaltz.r (hunk 1 above)
Comment (iraiosub) [reply modules/nf-core/ribowaltz/templates/ribowaltz.r]: done. Referencing commit 2c16f8e.

File: modules/nf-core/ribowaltz/templates/ribowaltz.r (hunk 3 above)
Comment (iraiosub) [reply modules/nf-core/ribowaltz/templates/ribowaltz.r]: thanks! done. Referencing commit 2c16f8e.

File: modules/nf-core/ribowaltz/templates/ribowaltz.r (hunk 4 above)
Comment (iraiosub) [reply modules/nf-core/ribowaltz/templates/ribowaltz.r]: deleted. Referencing commit 2c16f8e.

File: modules/nf-core/ribowaltz/templates/ribowaltz.r (hunk 5 above)
Comment (iraiosub) [reply modules/nf-core/ribowaltz/templates/ribowaltz.r]: moved up and added documentation for all functions. Referencing commit 2c16f8e.

File: modules/nf-core/ribowaltz/meta.yml (hunk 2 above)
Comment (iraiosub) [reply modules/nf-core/ribowaltz/meta.yml]: done, see commit 2c16f8e.

**Comment by SPPearce** [comment]
@pinin4fjords , do you want to review again?

**Comment by iraiosub** [comment]
@SPPearce thanks for the feedback - hopefully now all is resolved.

@pinin4fjords Likewise, many thanks! Let me know if there's more issues outstanding. 

Re the code additions to riboWaltz, I've kept this to a minimum. All the functions and code are simply calling riboWaltz functions and extracting and exporting the resulting tables and plots, which is not natively supported by the tool.  

- To make it more clear which functions are riboWaltz specific, I have now used them with a namespace qualified call e.g. `riboWaltz::psite`
- Wherever possible, the parameters I exposed have the same name as the argument names used in the riboWaltz package
- I exposed the parameters that should be relevant to most users, of course the community can decide if even more are needed
- added some comments to explain these params in commit 5dbeb80

**Review by pinin4fjords — APPROVED** [review]
I'm happy enough now

**Pull request merged**